### PR TITLE
feat: ping-2165 init chat from post

### PR DESCRIPTION
**Pull request by brm-stnd**

### CODE REVIEW

**Comment by bastian-purrer** [comment]
Image description   CodeRabbit



Currently reviewing new changes in this PR...

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between 1ea2bed0c34af33ebb20b6c405482bab3e53ad3a and 2b58116aac497a99bec5bfb7a2522377e055e4c6 commits.
</details>

<details>
<summary>Files selected (4)</summary>

* controllers/chat/index.js (2)
* controllers/chat/initChatFromPost.js (1)
* joi-validations/chat.validations.js (1)
* routes/chat.js (2)
</details>






---

**Review by bastian-purrer — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between 1ea2bed0c34af33ebb20b6c405482bab3e53ad3a and 2b58116aac497a99bec5bfb7a2522377e055e4c6 commits.
</details>

<details>
<summary>Files selected (4)</summary>

* controllers/chat/index.js (2)
* controllers/chat/initChatFromPost.js (1)
* joi-validations/chat.validations.js (1)
* routes/chat.js (2)
</details>





<details>
<summary>Files not summarized due to errors (4)</summary>

* joi-validations/chat.validations.js (nothing obtained from openai)
* routes/chat.js (nothing obtained from openai)
* controllers/chat/initChatFromPost.js (nothing obtained from openai)
* controllers/chat/index.js (nothing obtained from openai)

</details>


<details>
<summary>Files not reviewed due to errors (4)</summary>

* routes/chat.js (no response)
* controllers/chat/index.js (no response)
* controllers/chat/initChatFromPost.js (no response)
* joi-validations/chat.validations.js (no response)

</details>


<details>
<summary>Review comments generated (0)</summary>

* Review: 0
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

File: controllers/chat/initChatFromPost.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    const user = {
+      name: userModel?.username,
+      id: req.userId,
+      image: userModel?.profile_pic_path,
+      username: userModel?.username
+    };
+
+    await client.connectUser(user, req.token);
+
+    const channel = client.channel('messaging', `fromPost-${postId}`);
```
Comment (fajarism) [inline controllers/chat/initChatFromPost.js:34]: @brm-stnd if channel id only includes postId, then how can we differentiate chat initiated by user A or user B?
I think we should have another differentiation.
e.g. `fromPost-${postId}-${selfUserId}`

This way we can know that this is chat initiated by selfUserId and at the postId

File: controllers/chat/initChatFromPost.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            name: [userModel?.username, targetUserModel?.username].join(', ')
+          }
+        });
+      }
+    } catch (e) {
+      console.log(e);
+    }
+
+    const response = {
+      ...createdChannel
```
Comment (fajarism) [inline controllers/chat/initChatFromPost.js:56]: We should include better_channel_member for any latest channel
You can check on the `BetterSocialCore.chat.updateBetterChannelMember`

**Review by fajarism — CHANGES_REQUESTED** [review]

**Review by fajarism — APPROVED** [review]

File: routes/chat.js [hunk 3]
```diff
@@ -18,6 +18,7 @@ const {
   searchGif,
   listFeaturedGif,
   registerShareGif,
+  initChatFromPost
```
Comment (bastian-purrer) [inline routes/chat.js:21]: Image description   CodeRabbit

The import of `initChatFromPost` from the chat controller is correctly added. This function will be used in the new route handler.
```diff
+   initChatFromPost
```

File: routes/chat.js [hunk 4]
```diff
@@ -49,6 +50,12 @@ router.post('/init-chat', auth.isAuth, chatController.initChat);
 router.post('/init-chat-anonymous', auth.isAuthAnonim, chatController.initChatAnonymous);
 router.post('/move-to-anon', auth.isAuthAnonim, validate(ChatValidation.moveToAnon), moveToAnon);
 router.post('/move-to-sign', auth.isAuthV2, validate(ChatValidation.moveToSign), moveToSign);
+router.post(
+  '/init-chat-from-post',
+  auth.isAuth,
+  validate(ChatValidation.initChatFromPost),
+  initChatFromPost
+);
```
Comment (bastian-purrer) [inline routes/chat.js:58]: Image description   CodeRabbit

The addition of a new route `/init-chat-from-post` is correct. It's good to see that the route is protected by authentication middleware and uses validation middleware for request validation before passing control to the `initChatFromPost` function.
```diff
+ router.post(
+   '/init-chat-from-post',
+   auth.isAuth,
+   validate(ChatValidation.initChatFromPost),
+   initChatFromPost
+ );
```

File: controllers/chat/initChatFromPost.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        await channel.updatePartial({
+          set: {
+            channel_type: channelType,
+            name: [userModel?.username, targetUserModel?.username].join(', ')
+          }
+        });
+      }
+    } catch (e) {
+      console.log(e);
+    }
```
Comment (bastian-purrer) [inline controllers/chat/initChatFromPost.js:59]: Image description   CodeRabbit

This block of code is updating the channel after it has been created. If possible, consider setting these properties when creating the channel to reduce the number of operations and potential points of failure.

58:
Avoid using `console.log` for error logging in production code. Use a proper logging library that can provide more context about the error and handle log rotation.

**Review by bastian-purrer — COMMENTED** [review]
Image description   CodeRabbit

<details>
<summary>Commits</summary>
Files that changed from the base of the PR and between 2b58116aac497a99bec5bfb7a2522377e055e4c6 and d684fd0edbf24e11b592bf2d2af205a7ab5ab630 commits.
</details>

<details>
<summary>Files selected (4)</summary>

* controllers/chat/index.js (2)
* controllers/chat/initChatFromPost.js (1)
* joi-validations/chat.validations.js (1)
* routes/chat.js (2)
</details>







<details>
<summary>Files skipped from review due to trivial changes (2)</summary>

* controllers/chat/index.js
* joi-validations/chat.validations.js

</details>

<details>
<summary>Review comments generated (6)</summary>

* Review: 6
* LGTM: 0

</details>

---

<details>
<summary>Tips</summary>

### Chat with Image description  CodeRabbit Bot (`@coderabbitai`)
- Reply on review comments left by this bot to ask follow-up questions. A review comment is a comment on a diff or a file.
- Invite the bot into a review comment chain by tagging `@coderabbitai` in a reply.

### Code suggestions
- The bot may make code suggestions, but please review them carefully before committing since the line number ranges may be misaligned. 
- You can edit the comment made by the bot and manually tweak the suggestion if it is slightly off.

### Pausing incremental reviews
- Add `@coderabbitai: ignore` anywhere in the PR description to pause further reviews from the bot.

</details>

**Pull request merged**